File: 0104-maximum-depth-of-binary-tree/0104-maximum-depth-of-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int max = 0;
    int maxDepth(TreeNode* root) {
        traverse(root, 0);
        return max;
    }

    void traverse(TreeNode* curr, int depth) {
        if (curr != NULL) {
            if (depth + 1 > max) {
                max = depth + 1;
            }

            traverse(curr->left, depth + 1);
            traverse(curr->right, depth + 1);
        }
    }
};

File: 0104-maximum-depth-of-binary-tree/0104-maximum-depth-of-binary-tree.py
from collections import deque
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def maxDepth(self, root: Optional[TreeNode]) -> int:
        depth = 0

        # queue = deque([root])

        # while queue:
        #     depth += 1

        #     size = len(queue)

        #     for _ in range(size):
        #         node = queue.popleft()

        #         if node.left:
        #             queue.append(node.left)

        #         if node.right:
        #             queue.append(node.right)

        def traverse(curr, currDepth):
            nonlocal depth
            if not curr:
                return

            depth = max(currDepth, depth)

            traverse(curr.left, currDepth + 1)
            traverse(curr.right, currDepth + 1)

        traverse(root, 1)

        return depth